Fix partial update for Revision C & E

habibrehmansg opened the pull request:
This pull request includes several changes to improve the functionality and performance of the `TuringSmartScreenLib` library and modify the display logic in `Example2/Program.cs`. The most important changes include fixing a coordinate calculation bug, adding a render count for tracking, and modifying the display logic to handle different screen positions.

### Improvements to `TuringSmartScreenLib`:

* Fixed the coordinate calculation in `DisplayPartialBitmap` to correctly compute the position based on `(y + h) * Width + x` instead of `(x + h) * Width + y` (`TuringSmartScreenLib/RevisionC.cs`).
* Added a `renderCount` variable to track the number of times a bitmap is rendered and included this count in the header of the bitmap update command (`TuringSmartScreenLib/RevisionE.cs`). [[1]](diffhunk://#diff-0eca1f991f96071a065af188f5a594b2150bf6fedb753f48553c2ca3558da589R32-R33) [[2]](diffhunk://#diff-0eca1f991f96071a065af188f5a594b2150bf6fedb753f48553c2ca3558da589L265-R293)
* Changed the `DisplayPartialBitmap` method to return a boolean indicating whether the response requires a resend and updated the `DisplayBitmap` method to handle this correctly (`TuringSmartScreenLib/RevisionE.cs`). [[1]](diffhunk://#diff-0eca1f991f96071a065af188f5a594b2150bf6fedb753f48553c2ca3558da589L223-R233) [[2]](diffhunk://#diff-0eca1f991f96071a065af188f5a594b2150bf6fedb753f48553c2ca3558da589L305-R313)

### Modifications to display logic in `Example2/Program.cs`:

* Updated the display logic to show the buffer in all four corners of the screen instead of iterating through the screen height (`Example2/Program.cs`).

habibrehmansg commented:
So when you perform a partial bitmap update for RevisionC, it expects an incremental count flag. Without the count, the response on next CommandQuery will be "needReSend:1" and all subsequent bitmap updates will be ignored. If the partial update exceeds the boundaries of the screen (by some margin), "needReSend:1" will appear also.

I am not too sure how the device decides when to "needReSend:1" and hence decided to expose the flag instead for upstream logic to handle a full bitmap update as required. Feel free to update/refractor it accordingly.

RevisionE has a slightly different update structure, needing RGBA instead of RGB. However, I could not get the device to display any partial updates yet even without "needReSend:1" after each call. TLDR, the partial update functionality is not working, I might be missing something.

habibrehmansg commented:
I have discovered an issue with RevisionE where an exception is thrown "The semaphore timeout period has expired. : 'COM7'." and the device hangs (need to replug to fix) on exactly the 15th 1920x480 bitmap write. Restarting the program every 14th update does not trigger this issue.

Will update this PR if i can find a fix.

usausa commented:
@habibrehmansg 
I consider the following considerations necessary.

- I do not want to make CanDisplayPartialBitmap a public property and make users aware of it.
- I think that the process (retry, etc.) when needReSend is returned should be closed in each implementation.

habibrehmansg commented:
> @habibrehmansg 
> 
> I consider the following considerations necessary.
> 
> 
> 
> - I do not want to make CanDisplayPartialBitmap a public property and make users aware of it.
> 
> - I think that the process (retry, etc.) when needReSend is returned should be closed in each implementation.
> 
> 

How would you propose for the caller to be aware that the screen is rejecting partial updates, and therefore needs a full redraw? 

Should the full bitmap be kept and redrawn silently?

usausa commented:
I would also like to see what approach is taken in Python libraries.
https://github.com/mathoudebine/turing-smart-screen-python

Also, if it is unavoidable that a full update will be required when a partial update fails, I would consider having each implementation have a copy of Buffer for the screen size and use that to update the full screen.
However, this action will only be taken after the Python implementation has been checked and the case where the error occurs has been fully investigated.

I don't have time right now, but I will confirm this phenomenon as well.

usausa commented:
I have taken some of the PR ahead of time and imported it into RevisionC.
Has this improved the situation?
https://github.com/usausa/turing-smart-screen/commit/a86b57b931afc2e79380ac7bc4798df77f07d592

As far as I checked, I did not encounter any problems after repeated partial updates.

I would also like to consider setting the return value of DisplayBuffer/DisplayBitmap to bool instead of CanDisplayPartialBitmap as a means of knowing when a drawing has failed.

habibrehmansg commented:
> I have taken some of the PR ahead of time and imported it into RevisionC. Has this improved the situation? [a86b57b](https://github.com/usausa/turing-smart-screen/commit/a86b57b931afc2e79380ac7bc4798df77f07d592)
> 
> As far as I checked, I did not encounter any problems after repeated partial updates.
> 
> I would also like to consider setting the return value of DisplayBuffer/DisplayBitmap to bool instead of CanDisplayPartialBitmap as a means of knowing when a drawing has failed.

Yes, it works. The same fix would work for Rev E.

Setting to bool would also work to inform the caller. 

Also, update on Rev E,  port.DiscardInBuffer();port.DiscardOutBuffer(); is required otherwise 15 full updates will cause a port timeout and only fixed by replugging. I am not sure why.

usausa commented:
Since the return value has been changed to bool, make the same changes to the PR.
Example should remain unchanged.

Rev C also adds port.DiscardInBuffer() to ReadResponse().
Make the same changes in Rev E first and see if the problem still occurs.
If it improves, I would like to merge at one end there.

habibrehmansg commented:
> Since the return value has been changed to bool, make the same changes to the PR. Example should remain unchanged.
> 
> Rev C also adds port.DiscardInBuffer() to ReadResponse(). Make the same changes in Rev E first and see if the problem still occurs. If it improves, I would like to merge at one end there.

-Updated Rev E to be similar
-port.DiscardInBuffer() in ReadResponse() for Rev E results in port read timeout error immediately after on the next read attempt. Does not happen on Rev C.
-Fix a partial update position calculation bug on Rev C (and E). I know you asked to not change example, but I added in a litmus test to visually see that the position of partial update is calculated correctly.

usausa commented:
I would like to merge, but I would like to confirm one point.
Is DiscardOutBuffer always needed, not just when DisplayPartialBitmap fails?

I am planning to get Rev E, so I will check on the buffer myself when I get it.

habibrehmansg commented:
> I would like to merge, but I would like to confirm one point. Is DiscardOutBuffer always needed, not just when DisplayPartialBitmap fails?
> 
> I am planning to get Rev E, so I will check on the buffer myself when I get it.

I tested a loop of multiple full panel updates. Without DiscardOutBuffer, the panel would freeze occasionally resulting in timeout on read. The reverse is true for Rev C (hence I removed it). On multiple partial updates, it would freeze occasionally.

usausa commented:
If possible, could you also add a partial update loop to the Example, which is a problem when there is no DiscardOutBuffer?
I want to check it out when I get my own 8.8”.

habibrehmansg commented:
The example would not work for E. For some reason E does not accept the PNG loaded via SKBitmap. I'm not too familiar with it. Also, there would need to be a new background image of size 480x1920 for E. Loading the current one will not work either. Should there be an Example3?

usausa commented:
Ok, I'll check back later on the DiscardOutBuffer and merge it at one end.

usausa commented:
@habibrehmansg 
I welcome alpha support and video background support if I can get PR.
If alpha and video support is to be added to this library, the hope is that the policy will be as follows.

- interface does not change existing methods, but prepares another method for alpha(as much as possible)
- Consider implementing two separate IScreenBuffer implementations for displays that support alpha, such as ScreenBufferRgbX and ScreenBufferArgbX.
- IScreen.DisplayBuffer(), use different DisplayBitmap() to be used depending on the type of IScreenBuffer implementation
- Rev A and Rev B implementations delegate alpha to ignored processes

The following is an image.

```csharp
public interface IScreenBuffer : IDisposable
{
    void SetPixel(int x, int y, byte r, byte g, byte b);

    void SetPixel(int x, int y, byte a, byte r, byte g, byte b);
}

public sealed class TuringSmartScreenBufferRgbX : IScreenBuffer
{
}

public sealed class TuringSmartScreenBufferArgbX : IScreenBuffer
{
}

public sealed unsafe class TuringSmartScreenRevisionX
{
    public bool DisplayBitmap(int x, int y, byte[] bitmap, int width, int height);

    public bool DisplayAlphaBitmap(int x, int y, byte[] bitmap, int width, int height);
}

internal sealed class ScreenWrapperRevisionX
{
    public override bool DisplayBuffer(int x, int y, IScreenBuffer buffer) => 
        buffer is TuringSmartScreenBufferArgbX alphaBuffer
            ? screen.DisplayAlphaBitmap(x, y, alpha.Buffer, buffer.Width, buffer.Height);
            : screen.DisplayBitmap(x, y, ((TuringSmartScreenBufferX)buffer).Buffer, buffer.Width, buffer.Height);
}
```

For future refactoring, we will consider the following forms.

- For ScreenBuffer, create in units like Rgb24, Argb32, Rgb565 instead of per Rev
- Add the type Enum property to IScreenBuffer

For video, there will be a separate interface from IScreen.

```csharp
public interface IVideoScreen
{
    void Start();

    void Stop();

    void Upload(Stream stream);

    ...
}

internal sealed class ScreenWrapperRevisionX : ScreenBase, IVideoScreen
{
}
```

Would you consider it?

habibrehmansg commented:
@usausa 
Ok, thanks for the direction! I'll open a PR when I am done.

usausa commented:
I also got the 8.8 inch screen.
I was wrong about Alpha.
I think alpha should be handled in the IScreenBuffer.
I plan to do an update on the IScreenBuffer。

```csharp
public interface IScreenBuffer : IDisposable
{
    int Depth { get; }

    int Width { get; }

    int Height { get; }

    void SetPixel(int x, int y, byte r, byte g, byte b, byte a = 255);

    void Clear();

    void Clear(byte r, byte g, byte b, byte a = 255);
}

// For RevisionE BGRA buffer
public sealed class ScreenBufferBgra32 : IScreenBuffer
{
...
}
```
